src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1428,107 +1455,139 @@ private bool TryExecuteSymbolActions(SymbolDeclaredCompilationEvent symbolEvent,
+                    {
+                        if (!processedAnalyzers.Contains(analyzer))
+                        {
+                            AnalyzerExecutor.MarkSymbolEndAnalysisComplete(symbol, analyzer, analysisStateOpt);
+                            completedAnalyzers.Add(analyzer);
+                        }
+                    }
+                }
 
-                if (subsetProcessedAnalyzersBuilderOpt.Count > 0)
+                if (!success)
                 {
-                    subsetProcessedAnalyzers = subsetProcessedAnalyzersBuilderOpt.ToImmutableAndFree();
+                    Debug.Assert(completedAnalyzers.Count < analysisScope.Analyzers.Length);
+                    subsetProcessedAnalyzers = completedAnalyzers.ToImmutable();
+                    return false;
                 }
                 else
                 {
-                    subsetProcessedAnalyzersBuilderOpt.Free();
+                    subsetProcessedAnalyzers = ImmutableArray<DiagnosticAnalyzer>.Empty;
+                    return true;
                 }
             }
-
-            return success;
+            finally
+            {
+                processedAnalyzers.Free();
+                completedAnalyzers.Free();
+            }
         }
 
-        private static SyntaxNode GetTopmostNodeForAnalysis(ISymbol symbol, SyntaxReference syntaxReference, Compilation compilation)
+        private static SyntaxNode GetTopmostNodeForAnalysis(ISymbol symbol, SyntaxReference syntaxReference, Compilation compilation, CompilationData compilationData, CancellationToken cancellationToken)
         {
-            var model = compilation.GetSemanticModel(syntaxReference.SyntaxTree);
-            return model.GetTopmostNodeForDiagnosticAnalysis(symbol, syntaxReference.GetSyntax());
+            var model = compilationData.GetOrCreateCachedSemanticModel(syntaxReference.SyntaxTree, compilation, cancellationToken);
+            return model.GetTopmostNodeForDiagnosticAnalysis(symbol, syntaxReference.GetSyntax(cancellationToken));
         }
 
         protected abstract bool TryExecuteDeclaringReferenceActions(
             SymbolDeclaredCompilationEvent symbolEvent,
             AnalysisScope analysisScope,
             AnalysisState analysisStateOpt,
             bool isGeneratedCodeSymbol,
-            in AnalyzerActions additionalPerSymbolActions,
+            in IGroupedAnalyzerActions additionalPerSymbolActions,
             CancellationToken cancellationToken);
 
+        protected abstract IGroupedAnalyzerActions EmptyGroupedActions { get; }
+        protected abstract IGroupedAnalyzerActions CreateGroupedActions(DiagnosticAnalyzer analyzer, in AnalyzerActions analyzerActions);
+
+        protected interface IGroupedAnalyzerActions
+        {
+            bool IsEmpty { get; }
+            AnalyzerActions AnalyzerActions { get; }
+            IGroupedAnalyzerActions Append(in IGroupedAnalyzerActions groupedAnalyzerActions);
+        }
```

Comment:
1. extract to file.
2. no point in `in` afaict.

Reply:
Done, thanks!